### DISCUSSION TITLE
Fragmented list with no bullet/numbering

### DISCUSSION THREAD

**Discussion by sbibauw** [marp-team/marp#270] [Ideas] [answered]
The automatic *fragmented list* obtained with `*`  is simply, in my opinion, the killer feature of Marp.

```
* Appears first
* Appears second, after a click
* Appears after another click
```

Pure goodness! Extremely fast to write, pedagogically super powerful. 

Now, I sometimes would like to obtain the same fragmented list effect without the bullet point (or the numbering of `1) 2) 3)`).
For instance, because I want to make 3 small images appear one after the other (the bullet messes with the layout). Also because it's not really a list.

I know it's possible to add it manually with a lot of HTML (something like: `<ul style="list-style-type:none"><li data-marpit-fragment="1">One</li></ul>`) but the reason I'd like to use `*` is not so much because it's a list, but rather because it makes it _SO easy_ to have fragmented display.

So, I think it would be fantastic to have a third type of Fragmented list, obtained for instance with `+` (CommonMark compliant):

```
+ One
+ Two
+ Three
```

which would render as

```
<ul style="list-style-type:none">
    <li data-marpit-fragment="1">One</li>
    <li data-marpit-fragment="2">Two</li>
    <li data-marpit-fragment="3">Three</li>
  </ul>
```

(P.S. Only partially related, but I would also so much like to have automatic fragmented display of **tables**, either row by row, or column by column. Maybe with a header instruction such as: `<!-- table: row -->`. Otherwise, tables are often too complex/cluttered to present all at once, and using HTML here is simply impossible if I want to keep my sanity.)

**Comment by yhatt** [comment]
Unfortunately, Marp team has recognized the fragmented list as evil feature rather than the killer feature. We are in favor of the following presentation practice:

> No Builds. No presentation needs builds (exposing one bullet point of a list at a time), animations or transitions; no exceptions.
>
> _&mdash; [Presentation Rules – Jilles.net](http://www.jilles.net/perma/2020/06/05/presentation-rules.html)_

In addition, new syntax with implicit assigning `list-style-type: none` to `style` attribute would bring confusions users who know that `+` will render as the list in HTML.

We always recommend to make a Marp plugin, to transform `+` into the fragmented list with specific HTML. (You can be based on [Marpit fragment plugin](https://marpit-api.marp.app/markdown_fragment.js.html))

**Reply by sbibauw** [reply]
Thanks, Yuki.
I disagree with the "no builds" rule (or "no bullets", if I understand your reference well) as a universal, though. I do know best practices for presentations and I can understand why, in a corporate/business or dev context, no bullets, only 1 idea per slide with a title and a plot, or something like that, could make sense.
But that does not work well for concept-heavy, information-dense presentations as the ones we do in teaching and in academia. (Or do we maybe overestimate our students' ability to assimilate concepts and information? 🤔)
For instance, let's say you're introducing Asimov's [Three Laws of Robotics](https://en.wikipedia.org/wiki/Three_Laws_of_Robotics): you would want to have all three visible at the same time on one slide, as:
```
1) Robot must not injure human. 
2) Robot must obey human, except if contrary to (1). 
3) Robot must protect itself, except if contrary to (1) or (2).`
```
And here, builds are great. You really want to have first 1) showing alone, without any other, but you also want all 3 to be visible at the same time.

The only way it could be made even better, in my opinion, would be if (1) and (2) would be (automatically) slightly greyed out when (3) is introduced, to make sure the audience is paying attention to the new information.

**Reply by yhatt** [reply]
I can understand your opinion. But essentially it's important to know that _Marp's goal is in different premise_.

Original Marp ([yhatt/marp](https://github.com/yhatt/marp)) was made for conference and business purpose, not for educational. So Marp team is still aim to following the best practice for business presentation like the mentioned article. _We are not so interested in what are in out of the goal_, such as the fragmented list, HTML output. Current features implemented in Marp are just a bonus. Even if implemented features that we are not interested, we cannot promise whether it will keep maintained in Marp for a long time.

Fortunately Marp has flexibillity like custom CSS styling and plugin architecure. If you had some lacked features, just making your own style or plugin would be fullfill your purpose. Actually many users have turned over Marp to educational presentation by making own customize.

**Reply by sbibauw** [reply]
OK, good to know actually. Note just one thing though: you might, in the future, be interested to know who your users are (quick survey?). Marp's main "selling point" in my case is its speed: I can create content-intensive presentations very fast. As a university lecturer, I need to do ±10h of presentations per week and Marp is so powerful there. If I had to do only 1-5 presentations per year, I would probably use another tool (for developers, Reveal & Impress work pretty well; for business or corporate events, Powerpoint is well integrated with Excel and more customizable). But Marp is a game-changer (thanks to the fast builds, among other things) for academics and teachers who have to present a lot. So maybe there are more of us than you think? 😉

**Reply by yhatt** [reply]
If Marp had been created for commercial, taking action to resolve user needs is the best. But in reality, Marp is a personal project to resolve my pains, and just making public as an open source backed by volunteer work in weekend. Thus, I would not so work to not-interested features for me _even if there are a lot of user needs._ e.g. Presenter view: I'm not much work to it, but [a contributor in community is working to extend](https://github.com/marp-team/marp-cli/pulls?q=is%3Apr+author%3Achrisns)

In addition, maintaining many built-in features would lead to lose project simplicity and sustainability. As you have described, Marp is now making worth in various fields over our assumption. So we have to avoid [burning out this project](https://github.com/yhatt/marp/issues/174) again by [a lot of requests that are not suited to our goal](https://github.com/yhatt/marp/issues?q=is%3Aissue). :fire: 

> I'm just saying it is not suitable as a built-in feature, and note that I was never saying your suggestion is impossible. The open source ecosystem and [plugin architecture](https://marpit-api.marp.app/) should help to realize your think. :)

**Reply by Juanvvc** [reply]
I agree with yhatt: as much as I would like to see some features in Marp useful for the academia, this is an open source project and not a commercial solution. Marp is not driven by market statistics and the distribution of users or their needs is irrelevant. The force that is going to move the project one way or the other is the enthusiasm, time and interests of Marp developers, not Marp users.

I'm very grateful that the interests of the Marp team are aligned just by chance with my needs, and I can be a user of Marp with its current features.

Thanks, yhatt.

**Reply by sbibauw** [reply]
Thank you. You're totally right! Thanks a lot for all your hard work, Yuki. Indeed, we are really lucky our needs aligned with the product you created. Thanks again for Marp! 🙏

**Comment by Juanvvc** [comment, marked as answer]
You can change the CSS to suit your needs. In this case, add this to you slide and use the normal fragmented list:

```
# Hidden bullets

<style scoped>li { list-style-type: none}</style>

* These bullets
* are hidden
* to the viewer
```

More complex example, personalizing the CSS for each item individually: https://juanvvc.github.io/markdown-slides/marp/marp-tutorial.html#20  Code: https://github.com/Juanvvc/markdown-slides/blob/master/marp/marp-tutorial.md?plain=1#L372

You can also create CSS classes:

```
<style>.hidden-bullets li { list-style-type: none}</style>

# Normal slide

- Normal
* Fragmented
* List

# Hidden bullets
<!-- _class: hidden-bullets -->

- These bullets
* are hidden
* to the viewer
```

Also, notice: you can use `-` for the first item in the list, and the first item will appear from the beginning. I think this is a nice feature if you are planning to use animated lists.

**Reply by yhatt** [reply]
Yes. Twekable CSS is one of reason why we does not have `+` syntax :)

**Reply by sbibauw** [reply]
Thanks a lot, Juan! Great solution!

**Reply by bilogic** [reply]
Can a `# header` be fragmented too?

**Reply by yhatt** [reply]
@bilogic Please refer to [marp-team/marpit#195](https://github.com/marp-team/marpit/issues/195#issuecomment-534819826). You can be fragmented anything by wrapping with `<div data-marpit-fragment>`. (You may also require to set the preference to allow HTML, for using `data-*` attribute)

```markdown
<div data-marpit-fragment>

# Header

</div>
```

**Reply by bilogic** [reply]
Thanks, I tried and did allow HTML, but it garbled up the final HTML output by showing raw markdown

**Reply by yhatt** [reply]
It's important to make ensure separating the Markdown and HTML elements with a blank line. If not, raw markdown would be rendered as a part of HTML.

#### ✅ Correct

```markdown
<div data-marpit-fragment>

# Header

</div>
```

#### ❌ Incorrect

```markdown
<div data-marpit-fragment>
# Header
</div>
```
